### ISSUE TITLE
When I use the Decorators to change the function's return , pylint doesn't recognize it

### ISSUE THREAD

**Issue by 1997shp** [pylint-dev/pylint#3954]
### Steps to reproduce
1.
```
class Hook:
    @classmethod
    def request_reflect(func):
        @wraps(func)
        def wrapss(*args, **kwargs):
            data = {"a":"b"}
        
            return data
        return wrapss


class Main(Hook):

    def void_main():
        a = 1
        b = 2
        return a, b

main = Main()

data = main.void_main()
assert data["a"] = "b"
```

### Current behavior
When I use the hook function to change the return output ,Pylint will still throw an error in assert data["a"] = "b"
error is "Sequence index is not an int, slice, or instance with __index__"

### pylint --version output
pylint   --2.6.0

**Comment by hippo91** [comment]
@1997shp thanks for your report. However i'm unable to reproduce the issue, because the code in your snippet doesn't seem to be valid `python` code. Could you please give us a piece of python code that is valid and that yield the problematic message? Thanks in advance.

**Comment by 1997shp** [comment]
> @ 1997shp感谢您的报告。但是，我无法重现该问题，因为您的代码段中的代码似乎不是有效的`python`代码。您能否给我们提供一段有效且产生问题消息的python代码？提前致谢。

> 
@hippo91 thank you for your attention, I re uploaded the code and error screenshot, look forward your prompt reply
```
import types
from functools import wraps

def duration(func):

    @wraps(func)
    def wrapper(*args, **kwargs):
        test_dic = {
            "a" : args[0],
            "b" : args[1]
        }

        return test_dic

    return wrapper

@duration
def test_a(a, b):

    a = a
    b = b

    return a, b

dic = test_a(1, 2)

print(dic['a'])
```

![image](https://user-images.githubusercontent.com/49704057/100558103-d9551c80-32e7-11eb-8e7f-f3d3b10d36fd.png)

**Comment by hippo91** [comment]
@1997shp thanks. I can reproduce it. Here is a smaller snippet that trigger only the problematic message:
```
cat bug_pylint_3954.py
#pylint: disable=missing-module-docstring, missing-function-docstring
from functools import wraps

def duration(func):
    @wraps(func)
    def wrapper(*args):
        test_dic = {
            "a" : args[0],
            "b" : args[1]
        }
        return test_dic
    return wrapper

@duration
def test_a(first, second):
    first *= 3
    second *= 3
    print("Never pass here!")
    return first, second

dic = test_a(100, 2)

print(dic['a'])
```

Running python over this file gives:
```
python bug_pylint_3954.py
100
```
and pylint:
```
************* Module bug_pylint_3954
bug_pylint_3954.py:23:6: E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```

**Comment by 1997shp** [comment]
> @ 1997shp谢谢。我可以复制它。这是一个较小的代码段，仅触发有问题的消息：
> 
> ```
> cat bug_pylint_3954.py
> #pylint: disable=missing-module-docstring, missing-function-docstring
> from functools import wraps
> 
> def duration(func):
>     @wraps(func)
>     def wrapper(*args):
>         test_dic = {
>             "a" : args[0],
>             "b" : args[1]
>         }
>         return test_dic
>     return wrapper
> 
> @duration
> def test_a(first, second):
>     first *= 3
>     second *= 3
>     print("Never pass here!")
>     return first, second
> 
> dic = test_a(100, 2)
> 
> print(dic['a'])
> ```
> 
> 在此文件上运行python可得到：
> 
> ```
> python bug_pylint_3954.py
> 100
> ```
> 
> 和pylint：
> 
> ```
> ************* Module bug_pylint_3954
> bug_pylint_3954.py:23:6: E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
> ```

**Comment by kini** [comment]
Is this a duplicate of #2578?

**Comment by Pierre-Sassoulas** [comment]
Probably of #259 too :)

**Comment by darkclouder** [comment]
I have the same issue with `@overload`:

```python
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring,too-few-public-methods,no-member

from typing import Generic, overload, TypeVar


S = TypeVar("S")
T = TypeVar("T")


class MyClass(Generic[S]):
    def __init__(self, default: S):
        self.default = default

    @overload
    def or_default(self, maybe: None) -> S:
        ...

    @overload
    def or_default(self, maybe: T) -> T:
        ...

    def or_default(self, maybe = None):
        return maybe or self.default


my_tuple = (1, 2)
my_obj = MyClass(1)

print(my_tuple[my_obj.or_default()])
```

yields

```
29:6: E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```

**Comment by Pierre-Sassoulas** [comment]
A lot of similar issues were listed in https://github.com/pylint-dev/pylint/pull/8332

**Comment by jacobtylerwalls** [comment]
Duplicate of #2578
